### PR TITLE
Federation presubmit deploy/test lock

### PR DESCRIPTION
**Pull request by madhusudancs**
The design is described in detail at https://github.com/kubernetes/test-infra/issues/2259#issuecomment-294042246. There is one minor difference compared to what is described in the issue comment. I just got rid of using kubeconfig file during locking. GCS doesn't support any form of transactionality across objects. So it is not possible to ensure that the operations don't deadlock if we operate on two objects one after the other. This mechanism just uses a lock and barrier on the same file to ensure that the locks don't deadlock.

### CODE REVIEW

**Comment by madhusudancs** [comment]
/assign @csbell 
/assign @fejta

This is ready for review. PTAL.

/unassign

**Comment by fejta** [comment]
Tests are not passing

**Comment by madhusudancs** [comment]
@fejta vendoring problems. Fixing them.

**Comment by madhusudancs** [comment]
cc @perotinus for review as well.

**Comment by fejta** [comment]
/cc @krzyzacy 

Madhu I am not super keen on importing 50,000 lines of vendored code unless we need it, and it is unclear that we do.

What would you think about using boskos for this locking? Seems to do essentially the same thing: https://github.com/kubernetes/test-infra/tree/master/boskos

Alternatively what do you think about using either a) gsutil or b) raw http commands to access GCS rather than taking all these dependencies? The `append_result()` method in [boostrap.py](https://github.com/kubernetes/test-infra/blob/master/jenkins/bootstrap.py#L340) does a strongly consitent read-modify-write using this.

**Comment by madhusudancs** [comment]
> What would you think about using boskos for this locking? Seems to do essentially the same thing: https://github.com/kubernetes/test-infra/tree/master/boskos

By reading the doc, it wasn't clear to me if boskos could be used for concurrency control the way we need here. I will talk to @krzyzacy about it.

> Alternatively what do you think about using either a) gsutil or b) raw http commands to access GCS rather than taking all these dependencies? The append_result() method in boostrap.py does a strongly consitent read-modify-write using this.

I am very sure about one thing. I don't want to go down the gsutil route. It is ugly and maintainable. I started down that path and gave up. I looked at append_result() in bootstrap.py. It effectively does what I am trying to do here. As far as I understand, it only does an append. When we modify it enough to fit to what we need in this PR, my gut feeling is that, it looks very much like this or a lot more.

I really want to write this in a type-safe way. This is complex and can very soon become unmaintainable if we try to overcome that. I think this deserves an exception. Is there a reason why you are against vendoring 50K lines of code. After all, it is going to sit in the vendor directory.

**Comment by madhusudancs** [comment]
Btw, I forgot to mention. All these could be achieved using a Kubernetes API server as well, if we have one available for these tests always. But that involves vendoring Kubernetes client libraries that's 300K lines of code :)

File: kubetest/federation.go [hunk 1]
```diff
@@ -17,7 +17,36 @@ limitations under the License.
...
+	"time"
+
+	"cloud.google.com/go/storage"
+	"google.golang.org/api/googleapi"
+)
+
+const (
+	conditionNotMet = "conditionNotMet"
+
+	lockFile = "federated-clusters.lock"
```
Comment (perotinus) [inline kubetest/federation.go]: lockFileName

?

File: kubetest/federation.go [hunk 2]
```diff
@@ -27,3 +56,294 @@ func FedUp() error {
...
+func parseBucketObject(base, name string) (string, string) {
+	dir := strings.TrimPrefix(base, "gs://")
+	splits := strings.SplitN(dir, "/", 2)
+	if len(splits) == 1 {
+		return splits[0], name
+	}
+	return splits[0], path.Join(splits[1], name)
+}
+
+type gcsLock struct {
```
Comment (perotinus) [inline kubetest/federation.go]: Maybe a comment about how this works would be good here? A little note would make it easier for someone else (right now I'm thinking "me") to modify this later. I'm not sure I quite follow how the barriers and data are interacting, so a quick description might help.

File: kubetest/federation.go [hunk 3]
```diff
@@ -27,3 +56,294 @@ func FedUp() error {
...
+	bucket string
+	obj    string
+	ctx    context.Context
+	client *storage.Client
+}
+
+func newLock(save, lock string) (*gcsLock, error) {
```
Comment (perotinus) [inline kubetest/federation.go]: What are these parameters?

File: kubetest/federation.go [hunk 4]
```diff
@@ -27,3 +56,294 @@ func FedUp() error {
...
+	bucket, obj := parseBucketObject(save, lock)
+	return &gcsLock{
+		bucket: bucket,
+		obj:    obj,
+		ctx:    ctx,
+		client: client,
+	}, nil
+}
+
+func (l *gcsLock) lock(data string) error {
```
Comment (perotinus) [inline kubetest/federation.go]: Does the GCS client guarantee atomicity and thread safety?

File: kubetest/federation.go (hunk 2 above)
Comment (perotinus) [inline kubetest/federation.go]: Would this be worth moving into its own package? Not necessary, but having the differentiation between public and private methods might be useful from a readability stand point.

File: kubetest/federation.go [hunk 5]
```diff
@@ -27,3 +56,294 @@ func FedUp() error {
...
+			}
+		} else if err != nil {
+			return false, fmt.Errorf("failed to close the writer for %s/%s: %v", l.bucket, l.obj, err)
+		}
+		return true, nil
+	})
+}
+
+func poll(cond func() (bool, error)) error {
+	ok, err := cond()
```
Comment (perotinus) [inline kubetest/federation.go]: Does NewTicker not return a ticker that fires an initial burst? That's annoying.

File: kubetest/federation.go [hunk 6]
```diff
@@ -27,3 +56,294 @@ func FedUp() error {
...
+		buildNum := os.Getenv("BUILD_NUMBER")
+		err = l.lock(buildNum)
+		if err != nil {
+			return err
+		}
+
+		// Only restore .kube/config from previous --up, use the regular
+		// extraction strategy to restore version.
+		log.Printf("Load kubeconfig from %s", o.save)
+		loadKubeconfig(o.save)
```
Comment (perotinus) [inline kubetest/federation.go]: Where does it get stored?

**Comment by fejta** [comment]
If this code is still alive please rebase and reopen the pr, thanks!
/close

**Pull request closed without merging**